Add `to_arrow_device` function to cudf interop using nanoarrow

zeroshade opened the pull request:
## Description
Introduce new `to_arrow_device` and `to_arrow_schema` functions to utilize the `ArrowDeviceArray` structure for zero-copy passing of libcudf::table.

Add nanoarrow as a vendored lib and a script to update it.

Initial step towards addressing #14926

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

jrhemstad commented on cpp/include/cudf/interop.hpp:
```diff
@@ -162,6 +164,49 @@ std::shared_ptr<arrow::Scalar> to_arrow(cudf::scalar const& input,
+ * @note Copies will be performed in the cases where cudf differs from Arrow
+ * such as in the representation of bools (Arrow uses a bitmap, cudf uses 1-byte per value).
+ */
+struct ArrowDeviceArray to_arrow_device(
+  cudf::table& table,
```
suggestion: I believe this should be an r-value ref to make it clear this API is taking ownership of the data owned by the `table`. 

```suggestion
struct ArrowDeviceArray to_arrow_device(
  cudf::table&& table,
```

vyasr commented:
JFYI I'll be on vacation for the next week or so. Feel free to proceed without my input, just letting y'all know that I will catch up when I get back.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,586 @@
+      case cudf::type_id::DURATION_SECONDS:
+      case cudf::type_id::DURATION_MILLISECONDS:
+      case cudf::type_id::DURATION_MICROSECONDS:
+      case cudf::type_id::DURATION_NANOSECONDS: storage_type = NANOARROW_TYPE_DURATION; break;
```
Do you think I shouldn't separate the `to_arrow_schema` from the `to_arrow_device` and just do both in the same function? I figured it made sense to maintain the separation like we do elsewhere so that multiple instances with the same schema don't require allocating the schema object multiple times. 

Alternately I could require passing an `ArrowSchema` into the `to_arrow_device` function

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,586 @@
+  ArrowBuffer* buf = ArrowArrayBuffer(out, i);
+  buf->data        = reinterpret_cast<uint8_t*>(device_buf->data());
+  // we make a new unique_ptr and move to it in case there was a custom deleter
+  ArrowBufferSetAllocator(buf,
+                          ArrowBufferDeallocator(&device_buffer_finalize<T>,
+                                                 new std::unique_ptr<T>(std::move(device_buf))));
```
if `buf->data == nullptr` then what use would it be to set `buf->size_bytes` to anything?

Should I change this now? Or just wait for it to get fixed in nanoarrow / do you have a suggestion that would be better than trying to repurpose the buffer deallocator like this?

GregoryKimball commented:
@zeroshade thank you for building out this proposal. We will review and provide feedback on the draft. 🙏

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,586 @@
+  set_buffer(buf, 1, tmp.get());
+
+  NANOARROW_RETURN_NOT_OK(
+    ArrowArrayFinishBuilding(tmp.get(), NANOARROW_VALIDATION_LEVEL_NONE, nullptr));
```
Are we sure that `NANOARROW_VALIDATION_LEVEL_MINIMAL` doesn't attempt to access the data at all and won't choke on GPU pointers?

zeroshade commented:
Would it be possible to get some more reviews here? I've been attempting to keep this up to date but I haven't dug too much further into building this until I get some confirmation that everyone is okay with this approach and that it is worthwhile to continue in this pattern versus refactoring or changing the approach.

GregoryKimball commented:
Thank you @zeroshade for keeping this up to date. At the moment we are juggling GTC-related activities so please excuse the delay in reviews.

One piece of feedback from the team is that we would prefer to pull in the nanoarrow headers at build time rather than adding the file to the source tree.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+
+    auto status = cudf::type_dispatcher(
+      col->type(), detail::dispatch_to_arrow_device{}, *col, col->type().id(), stream, mr, child);
+    CUDF_EXPECTS(status == NANOARROW_OK, "type dispatcher failed to convert to ArrowArray");
```
Same question as above about error handling.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+  cudaEvent_t sync_event;
+};
+
+static void ArrowDeviceArrayRelease(struct ArrowArray* array)
```
OK I see that this is being assigned as the release callback for the final object, so I guess you're trying to hide it here and only make it callable via the pointer. I still think the static is redundant though given the namespacing.

zeroshade replied:
I think you're right that the `static` is likely redundant given the anonymous namespace, i'll remove it

vyasr reviewed (changes requested):
Apologies for the slow response. I came back from vacation to a long backlog and too much to juggle, and I didn't want to give this PR a cursory review. I spent a decent chunk of my review reading the nanoarrow API docs to make sure I understood the code here so I may have missed some things in the code, but overall I think this PR is a great start! The approach looks good and I think we should move forward.

I have a couple of bigger picture questions:
1. I'm wondering how we want this to interoperate with the existing arrow interop in libcudf. In order to get libcudf working more directly off the C Data interface, do you think it would be possible to rewrite our existing `to_arrow` as basically (pseudocode) `cudaMemcpy(to_arrow_device(table), DtoH)`?
2. Possibly also a question for @paleolimbot: is there any plan for working with scalars? Scalars are a bit of thorn in our side right now, especially in the Python/C++ interop layer because of how many edge cases run into issues with different pyarrow APIs. Do you foresee a way to rewrite our scalar interop in terms of the C Data interface/nanoarrow functions coming at some point? I would absolutely settle for a solution that boils down to "python scalar->python array->C++ array->C++ scalar" and in reverse (since I don't imagine there are imminent plans to add scalars directly to the interface), but a big part of why I wrote https://github.com/rapidsai/cudf/pull/14121 was because even that first step "python scalar->python array" was lossy and sometimes raised errors in pyarrow. I know there are some [nascent nanoarrow Python bindings](https://github.com/apache/arrow-nanoarrow/tree/main/python), so maybe that's what we want to move towards in the long term.

As per Greg's comment, I think it would be good to get the nanoarrow files at build time. We typically do that with CPM.cmake via rapids-cmake. I'm not sure how familiar you are with that approach, but there are some good examples in [rapids-cmake](https://github.com/rapidsai/rapids-cmake/tree/branch-24.04/rapids-cmake/cpm). The [rmm](https://github.com/rapidsai/rapids-cmake/tree/branch-24.04/rapids-cmake/cpm/rmm.cmake) and [gtest](https://github.com/rapidsai/rapids-cmake/tree/branch-24.04/rapids-cmake/cpm/gtest.cmake) are good starting points that don't have too many additional complexities baked in. That said, if that CMake is unfamiliar to you don't worry about it, we can change the vendoring strategies at a later point once the PR is more mature.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+  return ArrowSchemaSetType(out, NANOARROW_TYPE_STRING);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::list_view>(column_view input,
+                                                        cudf::type_id,
+                                                        column_metadata const& metadata,
+                                                        struct ArrowSchema* out);
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::dictionary32>(column_view input,
+                                                           cudf::type_id,
+                                                           column_metadata const& metadata,
+                                                           struct ArrowSchema* out);
```
Because I recursively use the `dispatch_to_arrow_type` inside of the struct specialization, I needed to forward declare these or else I get a "specialization declared after use" compile error.

vyasr replied:
Ah got it I forgot about the recursion, you're right.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+  return decimals_to_arrow<DeviceType>(input, out);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<bool>(column_view input,
+                                             cudf::type_id,
+                                             column_metadata const&,
+                                             struct ArrowSchema* out)
+{
+  return ArrowSchemaSetType(out, NANOARROW_TYPE_BOOL);
+}
```
for some reason I thought that bool wasn't `is_rep_layout_compatible`, but I just double checked and it is. So this is a good point, I'll push it to the default case.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+    out->dictionary);
+}
+
+template <typename T>
+static void device_buffer_finalize(ArrowBufferAllocator* allocator, uint8_t*, int64_t)
+{
+  auto* unique_buffer = reinterpret_cast<std::unique_ptr<T>*>(allocator->private_data);
+  delete unique_buffer;
+}
```
> So at the end of set_buffer ownership has changed hands into a unique_ptr whose scope is managed (manually) by arrow's allocator. Is that right? And I guess that means private_data really can be used to hold anything and that is an intentional part of the design, we're not just abusing the fact that it's a void * to stuff something else in there?

That's exactly the case :smile:

zeroshade replied:
> Why does it need to be static? Why is the scope/linkage of this function special?

It doesn't *need* to be static, it's just convenient to ensure that it isn't being exposed outside this file and allow more aggressive optimizations by the compiler if it desires to do so as we only use this as a function pointer for release callbacks. If you'd prefer it to not be marked static, that's perfectly fine.

vyasr replied:
As per the direction on [the other thread](https://github.com/rapidsai/cudf/pull/15047#discussion_r1508108642) I think we can remove the static since it's redundant in the anonymous namespace.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+namespace cudf {
+namespace detail {
+namespace {
+ArrowType id_to_arrow_type(cudf::type_id id)
```
Is there a particular file you'd like me to pull them into as a common space?

vyasr replied:
Given some of our other discussions, let's not bother with that in this PR. We can do it in the follow-up where we unify the host and device APIs.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+    tmp->null_count = column.null_count();
+
+    auto contents = column.release();
+    if (contents.null_mask) { set_buffer(contents.null_mask, 0, tmp.get()); }
```
I don't believe that nanoarrow has any constants/macros for these but I can add one into this file at least and use that constant

vyasr replied:
Ah thanks for pointing that out, that's an important consideration that I wasn't aware of because I don't know the union layout. I'm pro defining a constant in this file for now while we think about the broader question. I agree that it's tricky if you can't generalize across al types :thinking:

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+
+    NANOARROW_RETURN_NOT_OK(
+      ArrowArrayFinishBuilding(tmp.get(), NANOARROW_VALIDATION_LEVEL_MINIMAL, nullptr));
+    ArrowArrayMove(tmp.get(), out);
```
If we've ended up allocating anything (such as the unique_ptrs holding the buffer pointers or the copied values for a boolean conversion, etc) then anything still in the `UniqueArray` will be properly released via the release callback when it destructs if we error out without having performed the move.

It's so much about polluting `out` so much as ensuring that we don't leave any memory leaks if we error out early.

vyasr replied:
Got it, we're taking advantage of C++ semantics to avoid leaks here.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+};
+
+template <typename DeviceType>
+int unsupported_decimals_to_arrow(cudf::column& input,
```
I was trying to avoid changing or touching the existing `to_arrow` / `from_arrow` stuff in order to ensure no potential changes in behavior for those functions and that unification and refactoring could be done as a second pass afterwards. But I can look into doing this conversion.

vyasr replied:
Yup I agree. Let's do it in a follow-up.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<bool>(cudf::column& column,
```
In this case we *do* need the specialization because we have to call the `bools_to_mask` function to fetch the data instead of just being able to use the buffer as is because cudf uses 1 byte per bool and Arrow represents bools via a bitmap. 

This *could* be consolidated if desired, it would just require the extra branching if statement to call `bools_to_mask` if we're a bool array in the base default overload. that's all

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+  CUDF_FAIL("Unsupported type for to_arrow_device");
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::list_view>(cudf::column& column,
+                                                          cudf::type_id id,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource* mr,
+                                                          struct ArrowArray* out);
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::dictionary32>(cudf::column& column,
+                                                             cudf::type_id id,
+                                                             rmm::cuda_stream_view stream,
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             struct ArrowArray* out);
```
same response as the above. I need it because we use `dispatch_to_type` in the other definitions so i have to forward declare the specializations

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+    CUDF_FAIL("failed to initialize ArrowSchema children");
+  }
+
+  // what i would give for a zip iterator....
```
i guess the reason is that I didn't know it existed :smile: I'll look into using it haha

zeroshade replied:
thrust's zip_iterator doesn't implement `begin` or `end` so I can't use it for simplifying this with a range-based for loop. So there isn't much of a benefit here to using the zip iterator here as far as I can tell, at least not over the existing for-loop

vyasr replied:
Oh well, at least now you know it's there! I agree, it doesn't simplify much here. I wasn't initially sure how you intended to use a zip iterator, but now that you mention begin/end I see what you wanted to do.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+
+    auto status = cudf::type_dispatcher(
+      col.type(), detail::dispatch_to_arrow_type{}, col, col.type().id(), metadata[i], child);
+    CUDF_EXPECTS(status == NANOARROW_OK, "type dispatcher failed to determine column type");
```
Good point, i'll update that

zeroshade replied:
Ah, right now we're only using `ArrowErrorCode` (the integer) we aren't passing around an `ArrowError` to capture the error message. Since it looks like the existing code uses `CUDF_EXPECTS for error handling which throws, we could use `NANOARROW_THROW_NOT_OK` instead of `NANOARROW_RETURN_NOT_OK` and just have it throw on error like the existing code

vyasr replied:
Yup I think this makes sense.

vyasr commented:
Let me know when you'd like me to take another pass on this PR.

@paleolimbot once this PR is further along let's revisit the scalar question and see what we can come up with. The core question (which is broader than just this PR) is, if we want both cudf and libcudf to eventually be independent of libarrow by using the Arrow C data interface, we need a way to ingest all arrow scalars. If we only provide C++ APIs for ingesting arrow tables/arrays and remove the ability to directly convert libarrow scalars into libcudf scalars, that makes the API potentially a bit cumbersome, but at least it is technically feasible using public libarrow APIs: the current `cudf::from_arrow(arrow::Scalar)` implementation boils down to using public libarrow APIs to convert the scalar to a libarrow array, which is then tossed into a table on which we call `cudf::from_arrow(arrow::Table)`, so an external user could write the same function using public libarrow/libcudf APIs. However, the Python side is currently a bit more troublesome. From my experimentation there is no single pathway that will successfully convert all pyarrow scalars into pyarrow tables, leaving a C++ libarrow->libcudf scalar converter as the only currently viable option (which in turn means requiring libarrow).

Perhaps it'll end up that I'll just need to categorize those issues and ask for fixes in pyarrow. I'll look into trying to capture all the failure modes I observe and see if we can make some progress on those in parallel to this PR work.

zeroshade commented:
@vyasr This should be good for you to take another pass now.

* I've updated the cmake to use `FetchContent` to pull nanoarrow at build time
* I've shifted declarations and definitions based on the feedback
* removed the extraneous `struct` before `ArrowArray` and `ArrowSchema`
* Implemented the `dispatch_to_arrow_device` specializations for list/struct/dictionary types
* responded to most of your comments and waiting for responses

This still is only the `to_arrow_device` portion and I still haven't touched the existing `to_arrow` / `from_arrow` implementations yet in favor of that work being a future PR to avoid any potential to break things.

I also still need to add tests, which is going to be an undertaking in and of itself.

vyasr commented:
I'll take a look next week, thanks! I agree that we shouldn't touch the existing `to_arrow`/`from_arrow` implementations in this PR. We can do that in a follow-up. I could go either way on implementing a `from_arrow_device` in this PR. We can decide that later, but I would be totally fine if you decided that would also be best to do in a separate PR.

vyasr commented:
@paleolimbot it looks like my information is out of date! pyarrow 13 fixed the biggest issue that had previously concerned me in https://github.com/apache/arrow/pull/36162. That means we can do all our Python interop without dealing directly with scalars in the compiled components (see #15213), which makes it significantly more likely that we can rely solely on the C Data interface everywhere.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+
+struct dispatch_to_arrow_type {
+  template <typename T, CUDF_ENABLE_IF(not is_rep_layout_compatible<T>())>
+  int operator()(column_view, cudf::type_id, column_metadata const&, struct ArrowSchema*)
```
I'll follow-up on the other thread where I asked this. @zeroshade I think the main action item here is to drop the superfluous structs in all the declarations for brevity. Thanks!

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+                 cudf::type_id,
+                 rmm::cuda_stream_view,
+                 rmm::mr::device_memory_resource*,
+                 struct ArrowArray*)
```
You're right, I wasn't fully context-switching from C to C++ mode in this file given that we're using both styles of APIs. I haven't tested that specifically with nvcc but I'd expect it to work the same given that that's just the standard C->C++ translation of how C++ uses the struct keyword (as long as there are no virtual functions), and under the hood nvcc is just using the same behavior.

For the pointer, I get that the producing API is an arrow C API that will give us a pointer, I was just wondering if we should rewrite all the functions in this file to take that by reference instead. But I guess that doesn't really buy us anything; it'd just force us to dereference the pointer at every call site (which is more verbose).

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,590 @@
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             struct ArrowArray* out)
+{
+  CUDF_FAIL("Unsupported type for to_arrow_device");
```
Got it. Then yes no need to consolidate. I'll leave this thread open as a reminder to go back and fill out the other implementations.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,697 @@
+
+ArrowDeviceArray to_arrow_device(cudf::table&& table,
+                                 rmm::cuda_stream_view stream,
+                                 rmm::mr::device_memory_resource* mr)
```
I don't think this function should take an mr. IIUC the only allocations that occur anywhere in this function are of temporaries, such as the bitmask when handling bool columns or the intermediate buffer for decimal types <128 bit. Please correct me if I am wrong, but if not then you should use rmm's current memory resource instead and remove this parameter. See the "Memory Allocation"/"Temporary Memory" section of [the developer guide](https://docs.rapids.ai/api/libcudf/stable/developer_guide).

zeroshade replied:
Those aren't temporaries though. The memory allocated for the bitmask for handling bool columns and the buffer for decimal types are what is returned at the end, they aren't used as intermediaries or temporaries.

In the header, this parameter is defaulted to the current memory resource which is what the developer guide says to do

vyasr replied:
Yikes you're totally right. That's my bad, those are definitely outputs. This looks right then, carry on.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,697 @@
+  ArrowArrayMove(tmp.get(), &private_data->parent);
+  struct ArrowDeviceArray result;
+  result.device_id = rmm::get_current_cuda_device().value();
+  // can/should we check whether the memory is managed/cuda_host memory?
```
For now I'd say not. There isn't really a good way to check this.

vyasr reviewed (changes requested):
This is looking great! Once my current comments are addressed I think it's ready to open up for more reviews. I think this PR is a good size, we can do `from_arrow_device` in a separate one.

Thinking forward to how we could unify the existing host `to_arrow` method with this one, how should ownership work? Right now this function is written such that the produced arrow struct has sole ownership of the data. In the cases we've considered up until now, that makes sense because libcudf's data structure aren't designed to share data so our only real option is to give up ownership to arrow. However, the conversion to host is a special case because what we typically want is for the input cudf objects to remain valid after the host conversion is completed. I had been envisioning that `to_arrow` would become roughly as simple as `copyDeviceToHost(to_arrow_device(...))`, but for that to be possible we would need a way for `to_arrow_device` to support creating these structs without invalidating the input columns. That would presumably be internal-only functionality since it would be dangerous to expose that kind of API, but I think we'll need that to really unify the code paths. Does that seem plausible to you? Am I missing anything?

vyasr commented:
/ok to test

zeroshade commented on cpp/include/cudf/interop.hpp:
At the moment it's due to an issue in how nanoarrow has implemented some things and where the `nanoarrow_device.h` is located. @paleolimbot can explain more in depth, but eventually we should be able to remove these declarations and instead just include `<nanoarrow/nanoarrow_device.h>` (correct me if i'm wrong Dewey!)

vyasr replied:
OK cool, so should we stop relying on nanoarrow headers for the host structs and copy those in as well? Right now we've only got the device structs copied.

zeroshade replied:
We're using the nanoarrow headers for nanoarrow itself at the moment, which already include the non-device structs. The only benefit we'd have to putting the host structs into this header would be that we would potentially be able to eliminate nanoarrow entirely from the exported header file.

I thought it was more convenient to simply use the definitions which were already in the nanoarrow headers instead, but I'm fine with copying the definitions directly into here instead if preferred.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,697 @@
+namespace cudf {
+namespace detail {
+namespace {
+static constexpr int kValidityBufferIdx       = 0;
+static constexpr int kFixedWidthDataBufferIdx = 1;
```
it's just a common c++-ism that ultimately comes from common use of Hungarian notation. In Hungarian, the word for "constant" is "konstans" so the k-prefix became a bit of wordplay and a common way to denote global constants. I can remove if it you prefer, at this point it's pretty much a habit for me as I prefer to ensure any global constant is denoted by the k-prefix so i can tell whenever I see the variable in use.

vyasr replied:
The last time I used Hungarian notation was in an all-PHP project :joy:

Yeah, I'd say let's remove this in favor of simple snake-cased constants (if you `git grep` the code base you'll see that's typically what we do, although it's not specified in our dev guide).

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,697 @@
+    contents.children[cudf::lists_column_view::offsets_column_index]->release();
+  NANOARROW_RETURN_NOT_OK(set_buffer(std::move(offsets_contents.data), 1, tmp.get()));
+
+  std::unique_ptr<cudf::column>& child =
```
I originally put the full type there instead of `auto` since it's not obvious that is the type on the right-hand side. Not a problem though, I'll switch it back to `auto`

zeroshade commented:
@vyasr I think i've addressed all of your comments (or replied to them) let me know what you think and if I should mark this as `Ready for review` :smile:

vyasr commented on cpp/CMakeLists.txt:
```diff
@@ -229,6 +236,9 @@ if(NOT BUILD_SHARED_LIBS)
   endif()
 endif()
 
+# need to add nanoarrow to the export set
+install(TARGETS nanoarrow EXPORT cudf-exports)
+
```
This should be superfluous now with the use of `rapids_cpm_find`

zeroshade replied:
Turns out it's not, at least I couldn't figure out how to make it unnecessary. I still get an error about `nanoarrow` not being in the export set unless I have this line. If you have any suggestions about how I can modify the `rapids_cpm_find` invocation to fix that, I'll give it a try.

vyasr commented on cpp/CMakeLists.txt:
```diff
@@ -730,6 +739,7 @@ target_include_directories(
          "$<BUILD_INTERFACE:${CUDF_SOURCE_DIR}/include>"
          "$<BUILD_INTERFACE:${CUDF_GENERATED_INCLUDE_DIR}/include>"
   PRIVATE "$<BUILD_INTERFACE:${CUDF_SOURCE_DIR}/src>"
+          "$<BUILD_INTERFACE:${nanoarrow_SOURCE_DIR}/src>"
```
@zeroshade I see that you removed this line, were you intending to hold off merging this PR until we get the necessary fixes in nanoarrow?

zeroshade replied:
Turned out that I didn't need this anymore after switching to `rapids_cpm_find`

vyasr replied:
This was indeed an issue in nanoarrow's CMake. I've resolved it in https://github.com/apache/arrow-nanoarrow/pull/406.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,681 @@
+    NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), storage_type, column));
+
+    auto contents = column.release();
+    if (contents.null_mask) {
+      NANOARROW_RETURN_NOT_OK(
+        set_buffer(std::move(contents.null_mask), kValidityBufferIdx, tmp.get()));
+    }
```
Just raising this again as an options in case you missed the suggestion, but totally fine if you intentionally decided not to do it: this could also go into `initialize_array` since the bitmask handling is the same for all types AFAICT (whereas the data and children handling is not).

zeroshade replied:
I intentionally chose not to do it because of the complexities of dealing with the call to `release` versus needing the size / null_count / etc. from the column *before* calling release. I mostly didn't want to juggle the handling and lifetimes

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,681 @@
+  ArrowSchemaInit(result.get());
+  NANOARROW_THROW_NOT_OK(ArrowSchemaSetTypeStruct(result.get(), input.num_columns()));
+
+  // what i would give for a zip iterator....
```
Let's remove this at this point :)

vyasr reviewed (changes requested):
A few small things, but yes I think this can be marked as ready for review now! I'll also trigger another CI run to see if we can start seeing this in action. I would recommend that you start working on tests now, that way those will be around when the next reviewer starts to take a look.

I'd appreciate if you could weigh in on my [overall question from my last review](https://github.com/rapidsai/cudf/pull/15047#pullrequestreview-1915265996). I think it would be quite helpful for us to have a plan for that before this PR is finished rather than having to bolt on changes later if we realize this design isn't general enough for that kind of us.

vyasr commented:
/ok to test

davidwendt commented on cpp/include/cudf/interop.hpp:
None of these are needed by function declarations or for the callers of these functions -- they are only used in internal source code. Perhaps they can be moved to the .cu file. I think we only need one of them so there is no need to define them all here. Another alternative is to move these to a `cudf/include/interop/detail/arrow.hpp` header instead.

I know these will be replaced with an include in the future but considering this is public header, everything here will be processed and published in our public documentation. I'd like to minimize the noise there as much as possible.

zeroshade replied:
Added a `cudf/include/interop/detail/arrow.hpp` header as suggested and moved this to it

davidwendt commented on cpp/include/cudf/interop.hpp:
```diff
@@ -162,6 +221,49 @@ std::shared_ptr<arrow::Scalar> to_arrow(cudf::scalar const& input,
+ *
+ * Populates and returns an ArrowSchema C struct using a table and metadata.
+ *
+ * @param input table_view to create a schema from
+ * @param metadata Contains the hierarchy of names of columns and children
+ * @return ArrowSchema generated from `input`
+ *
+ * @note For decimals, since the precision is not stored for them in libcudf,
+ * it will be converted to an Arrow decimal128 that has the widest precision the cudf decimal type
+ * supports. For example, numeric::decimal32 will be converted to Arrow decimal128 of the precision
+ * 9 which is the maximum precision for 32-bit types. Similarly, numeric::decimal128 will be
+ * converted to Arrow decimal128 of the precision 38.
```
```suggestion
 * @note For decimals, since the precision is not stored for them in libcudf,
 * decimals will be converted to an Arrow decimal128 which has the widest precision that cudf decimal type
 * supports. For example, `numeric::decimal32` will be converted to Arrow decimal128 with the precision of
 * 9 which is the maximum precision for 32-bit types. Similarly, `numeric::decimal128` will be
 * converted to Arrow decimal128 with the precision of 38.
 *
 * @param input table_view to create a schema from
 * @param metadata Contains the hierarchy of names of columns and children
 * @return ArrowSchema generated from `input` 
```

davidwendt commented on cpp/include/cudf/interop.hpp:
```diff
@@ -162,6 +221,49 @@ std::shared_ptr<arrow::Scalar> to_arrow(cudf::scalar const& input,
+ *
+ * @note Copies will be performed in the cases where cudf differs from Arrow
+ * such as in the representation of bools (Arrow uses a bitmap, cudf uses 1-byte per value).
+ */
```
You'll need to add `@param` tags here for each parameter to pass the style checks.

zeroshade replied:
didn't even realize i missed those! I've added them

davidwendt commented on cpp/include/cudf/interop.hpp:
```diff
@@ -162,6 +221,49 @@ std::shared_ptr<arrow::Scalar> to_arrow(cudf::scalar const& input,
+ * @note Copies will be performed in the cases where cudf differs from Arrow
+ * such as in the representation of bools (Arrow uses a bitmap, cudf uses 1-byte per value).
+ */
+ArrowDeviceArray to_arrow_device(
```
Perhaps this should return a `std::unique_ptr<ArrowDeviceArray>`
My understanding is the compiler is not required to perform [copy elision ](https://en.cppreference.com/w/cpp/language/copy_elision)so it may be safer to return a smart pointer instead.

davidwendt reviewed (changes requested):
This is looking good.

davidwendt commented:
/ok to test

zeroshade commented:
> However, the conversion to host is a special case because what we typically want is for the input cudf objects to remain valid after the host conversion is completed. I had been envisioning that to_arrow would become roughly as simple as copyDeviceToHost(to_arrow_device(...)), but for that to be possible we would need a way for to_arrow_device to support creating these structs without invalidating the input columns. That would presumably be internal-only functionality since it would be dangerous to expose that kind of API, but I think we'll need that to really unify the code paths. Does that seem plausible to you? Am I missing anything?

Yea, unfortunately we're going to need to create another, separate version of `to_arrow_device` in that case. Either we create a version that copies everything to the host as part of generating the `ArrowDeviceArray` struct, or we create a version that doesn't take any ownership or release the input and then create a `copyDeviceToHost` wrapper around it.

davidwendt commented:
/ok to test

davidwendt commented:
/ok to test

davidwendt commented on cpp/include/cudf/interop.hpp:
```diff
@@ -31,6 +31,7 @@
 
 #include <cudf/column/column.hpp>
 #include <cudf/detail/transform.hpp>
+#include <cudf/interop/detail/arrow.hpp>
```
I think this include can be moved to `to_arrow_device.cu`
And you should only need to forward reference the `ArrowDeviceArray` in this file by adding the following line before the `namespace cudf {` line:
```
struct ArrowDeviceArray;
```

zeroshade replied:
sure, I can clean this up with the forward declaration

davidwendt commented on cpp/include/cudf/interop/detail/arrow.hpp:
```suggestion
```
Do all of these need to be defined?

zeroshade replied:
since we'd no longer be requiring this via including it in `interop.hpp` I'm fine with removing the unnecessary definitions that aren't utilized by `to_arrow_device.cu`.

vyasr replied:
I think we're saying the same thing, but just to be certain: if we include any definitions of the Arrow C Data Interface in public headers we need to make sure that we copy everything including bits that we're not using because the preprocessor guards will ensure that we only have a single version of everything and we need to make sure that every TU that is being built gets what it need.

zeroshade replied:
@vyasr yea, we're saying the same thing. Since i'm using forward declarations in the public interface now, the internal `interop/detail/arrow.hpp` file doesn't need to have *everything* as it's not intended for public inclusion.

davidwendt commented:
/ok to test

davidwendt commented on cpp/include/cudf/interop/detail/arrow.hpp:
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2020-2024, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2024, NVIDIA CORPORATION.
```

davidwendt commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,681 @@
+/*
+ * Copyright (c) 2020-2024, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2024, NVIDIA CORPORATION.
```

davidwendt commented:
Should this PR include some gtests for this new API? Sorry if this has been already discussed.

zeroshade commented:
@davidwendt I'm starting to work on the gtests now

bdice commented on cpp/CMakeLists.txt:
```diff
@@ -209,11 +209,19 @@ include(cmake/thirdparty/get_fmt.cmake)
 # find spdlog
 include(cmake/thirdparty/get_spdlog.cmake)
 
+rapids_cpm_find(
```
Could we move this `rapids_cpm_find` call into a file like `cmake/thirdparty/get_nanoarrow.cmake`, and then include it? See the examples in that directory or ask if you need help.

bdice replied:
Something like this:

<details><summary>get_nanoarrow.cmake</summary>

```
# =============================================================================
# Copyright (c) 2024, NVIDIA CORPORATION.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License
# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
# or implied. See the License for the specific language governing permissions and limitations under
# the License.
# =============================================================================

function(find_and_configure_nanoarrow)
  set(oneValueArgs VERSION FORK PINNED_TAG)
  cmake_parse_arguments(PKG "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})

  rapids_cpm_find(
    nanoarrow ${PKG_VERSION}
    BUILD_EXPORT_SET cudf-exports
    INSTALL_EXPORT_SET cudf-exports
    CPM_ARGS
    GIT_REPOSITORY https://github.com/${PKG_FORK}/arrow-nanoarrow.git
    GIT_TAG ${PKG_PINNED_TAG}
    GIT_SHALLOW TRUE
  )
endfunction()

find_and_configure_nanoarrow(
  VERSION 0.4.0 FORK apache PINNED_TAG c97720003ff863b81805bcdb9f7c91306ab6b6a8
)
```

</details>

bdice replied:
Note that I defined export sets in the proposal above. I think that should work instead of manually exporting?

zeroshade replied:
even with those defined export sets, if I don't manually export I get 

```
CMake Error: install(EXPORT "cudf-exports" ...) includes target "cudf" which requires target "nanoarrow" that is not in any export set.
CMake Error in CMakeLists.txt:
  export called with target "cudf" which requires target "nanoarrow" that is
  not in any export set.
```

vyasr replied:
I've pushed something that should fix this for now. We can discuss in more depth down below since there are a few questions to be addressed about proper usage of nanoarrow I think.

zeroshade commented:
@davidwendt @bdice @vyasr I need a little bit of assistance if anyone has a moment. I'm trying to put the tests together but I'm not `cmake` savvy enough to figure out how and where to add the nanoarrow dependency to the test configurations. It's failing on finding the nanoarrow include directory and can't find the nanoarrow headers.

Thanks for any help you can provide!

vyasr commented:
@zeroshade I'll look into the build issues here, I should be able to get back to you by EOD tomorrow.

vyasr commented:
/ok to test

vyasr commented:
@zeroshade I've pushed some changes that should be sufficient to get things to build. However, I'm not entirely sure what you need in the tests. The data structure declarations should be available now via cudf (although there are some questions that we should address around the namespacing of the device struct, and I think we probably want to set `NANOARROW_NAMESPACE cudf`; right now all the structs are presumably available at the top-level namespace and can be used unqualified). The structs should be sufficient, right? In the test you'll be building cudf objects and then calling this API, so as long as you can declare the return type you'll be set.

There are some bigger questions about how best to build against with nanoarrow but that's probably better suited to a nanoarrow issue.

vyasr commented:
/ok to test

zeroshade commented:
@vyasr the issue i'm having with building the tests is that I can't figure out where to modify the `CMakeLists.txt` in the tests directory so that it can find the nanoarrow include directory. It's currently failing during compilation because it can't find `<nanoarrow/nanoarrow.hpp>`

vyasr commented:
Oh I see, I saw that there were a few issues with the build itself and I thought you might have been running into those as well. I'd actually be curious how you were getting things to build without my changes. Perhaps you weren't cleaning the build directory and you already had a local copy of nanoarrow so it was never executing some of the newer CMake code like the `rapids_cpm_find` after you made those changes from the original FetchContent? Even FetchContent doesn't support arbitrary commit hashes with shallow clones, see [this line from the docs](https://cmake.org/cmake/help/latest/module/ExternalProject.html#git)
> If GIT_SHALLOW is enabled then GIT_TAG works only with branch names and tags. A commit hash is not allowed.

so maybe one of the original copies of your code didn't do a shallow clone?

The reason you're running into problems is that nanoarrow headers are not part of the cudf target's public interface right now. That is by design; we're privately linking to nanoarrow (I'm talking about linking in CMake, not just compilation/linking) so that we don't export that dependency to consumers. nanoarrow is basically just a helper for us at build time, but consumers shouldn't need anything other than the definitions from the Arrow C Data Interface to use libcudf. The problem with including nanoarrow in our public link interface (which would allow the tests to find it) is that if someone tried to build another project against cudf, and if that project also leveraged nanoarrow internally, we would potentially run into conflicting headers (or worse, ODR violations if we also made the nanoarrow library linkage public and symbols leaked).

The tl;dr here is that, going back to [this earlier discussion](https://github.com/rapidsai/cudf/pull/15047#discussion_r1513703605), I think this is a clear indication that we should copy-paste the host structs into detail/interop.hpp and remove the include so that consumers of cudf headers don't transitively try to pull nanoarrow headers. If tests actually need nanoarrow on their own, I would link test targets directly to nanoarrow. I don't know if that's the case right now, but it shouldn't be a problem if it is since tests are built within the same CMakeLists.txt file so the target will be available (and if it weren't, you could include `get_nanoarrow.cmake` again).

vyasr commented:
> Yea, unfortunately we're going to need to create another, separate version of `to_arrow_device` in that case. Either we create a version that copies everything to the host as part of generating the `ArrowDeviceArray` struct, or we create a version that doesn't take any ownership or release the input and then create a `copyDeviceToHost` wrapper around it.

The latter is more what I was expecting. I am hoping we could use some relatively simple template magic where we rewrite some of our internal functions to accept forwarding references and then use something like [`is_rvalue_reference`](https://en.cppreference.com/w/cpp/types/is_rvalue_reference) to do some conditional branching around whether cudf releases the memory and whether or not to add a nontrivial release callback to the arrow struct. But at this stage that's still just a high-level concept, I haven't thought through the full implications of it yet.

davidwendt commented:
/ok to test

davidwendt commented:
/ok to test

davidwendt commented:
You should be able to run the style check locally. In my local environment I do 
```
git add --all .
pre-commit
```
The pre-commit will install all the dependencies it needs (only once) and then verifies any files staged for commit.
You can also run `pre-commit run --all-files` locally. This will run against the entire repo though does not usually take too long.

zeroshade commented:
@davidwendt i ran pre-commit, should be good now. i hope

davidwendt commented:
/ok to test

davidwendt commented:
Looks like you accidently pushed some `pyproject.toml` files

davidwendt commented:
/ok to test

zeroshade commented:
@davidwendt @vyasr Looks like the issue is on 
![image](https://github.com/rapidsai/cudf/assets/555095/6d943cbc-6456-4d1b-bea1-7e9029b467d2)

which isn't related to anything I've touched. Can I get some help on this one?

davidwendt commented:
Sorry, it looks like a bug that should be fixed once https://github.com/rapidsai/cudf/pull/15287 merges.

davidwendt commented:
/ok to test

davidwendt commented:
/ok to test

vyasr commented:
Is this waiting on a review from me now?

davidwendt commented:
> Is this waiting on a review from me now? Lost track a bit here sorry.

I was hoping you could debug the failing docs build.

vyasr commented:
I think It's not happy because ArrowDeviceArray is part of our public API as the return value of `to_arrow_device` but it is not documented. Adding a doxygen doc block to the declaration in interop.hpp should fix that (no need for much detail, best to probably just link to [the arrow docs](https://arrow.apache.org/docs/format/CDeviceDataInterface.html)).

jrhemstad commented on cpp/include/cudf/interop.hpp:
```diff
@@ -162,6 +164,54 @@ std::shared_ptr<arrow::Scalar> to_arrow(cudf::scalar const& input,
+ * @return ArrowSchema generated from `input`
+ */
+std::unique_ptr<ArrowSchema> to_arrow_schema(cudf::table_view const& input,
+                                             std::vector<column_metadata> const& metadata);
```
Any reason this can't be a span instead? `vector const&` is kind of an anti-pattern. 

```suggestion
                                             std::span<column_metadata const> metadata);
```

zeroshade replied:
Mostly i was just following the existing pattern (look at `to_arrow` which already uses `vector const&`).  I'm not opposed to it being a span, but personally i prefer consistency

zeroshade replied:
does libcudf build with C++20? I tried changing this to std::span but I'm getting compiler errors that it can't find `std::span` even with including `<span>`

bdice replied:
We require C++17 but have a span class for this purpose. Use cudf::host_span like this: https://github.com/rapidsai/cudf/blob/933e32ab9ad8e5057282c48129ddbd745c538967/cpp/include/cudf/concatenate.hpp#L94

vyasr commented:
/ok to test

vyasr commented:
/ok to test

zeroshade commented:
@vyasr Do i need to shift the doxygen block to the `interop/detail/arrow.hpp` file? Because even with the block I added to `interop.hpp` it's still yelling about missing the ArrowDeviceArray identifier.

vyasr commented:
Hmm let me take a look.

vyasr commented:
@zeroshade I believe the issue is that it's not in a group. Try adding `@ingroup copy_gather` below the `@brief` for these.

zeroshade commented:
@vyasr okie doke, I gave it a try and added the `@ingroup` lines. Let's see what happens

bdice commented:
/ok to test

zeroshade commented:
@vyasr Looks like we got the same error. Can you give it a shot?

vyasr commented:
Yeah let me take a look. If that didn't work, I'm guessing this has to do with the namespacing. One thing that I didn't notice before is that this is only happening to the ArrowDeviceArray but not to the ArrowSchema. That seems likely to be an important indicator.

davidwendt commented on cpp/tests/interop/to_arrow_device_test.cpp:
```diff
@@ -0,0 +1,735 @@
+
+  auto data_ptr        = input.get_column(0).view().data<int64_t>();
+  auto got_arrow_array = cudf::to_arrow_device(std::move(input));
+  ASSERT_EQ(rmm::get_current_cuda_device().value(), got_arrow_array->device_id);
+  ASSERT_EQ(ARROW_DEVICE_CUDA, got_arrow_array->device_type);
+
+  ASSERT_EQ(data.size(), got_arrow_array->array.length);
+  ASSERT_EQ(0, got_arrow_array->array.null_count);
+  ASSERT_EQ(0, got_arrow_array->array.offset);
+  ASSERT_EQ(1, got_arrow_array->array.n_children);
+  ASSERT_EQ(nullptr, got_arrow_array->array.buffers[0]);
+
+  ASSERT_EQ(data.size(), got_arrow_array->array.children[0]->length);
+  ASSERT_EQ(0, got_arrow_array->array.children[0]->null_count);
+  ASSERT_EQ(0, got_arrow_array->array.children[0]->offset);
+  ASSERT_EQ(nullptr, got_arrow_array->array.children[0]->buffers[0]);
+  ASSERT_EQ(data_ptr, got_arrow_array->array.children[0]->buffers[1]);
+
+  ArrowArrayRelease(&got_arrow_array->array);
```
```suggestion
  EXPECT_EQ(rmm::get_current_cuda_device().value(), got_arrow_array->device_id);
  EXPECT_EQ(ARROW_DEVICE_CUDA, got_arrow_array->device_type);

  EXPECT_EQ(data.size(), got_arrow_array->array.length);
  EXPECT_EQ(0, got_arrow_array->array.null_count);
  EXPECT_EQ(0, got_arrow_array->array.offset);
  EXPECT_EQ(1, got_arrow_array->array.n_children);
  EXPECT_EQ(nullptr, got_arrow_array->array.buffers[0]);

  EXPECT_EQ(data.size(), got_arrow_array->array.children[0]->length);
  EXPECT_EQ(0, got_arrow_array->array.children[0]->null_count);
  EXPECT_EQ(0, got_arrow_array->array.children[0]->offset);
  EXPECT_EQ(nullptr, got_arrow_array->array.children[0]->buffers[0]);
  EXPECT_EQ(data_ptr, got_arrow_array->array.children[0]->buffers[1]);

  ArrowArrayRelease(&got_arrow_array->array);
```
I'm not sure if these warrant an assert. I'm a bit worried that triggering an assert would cause a memory leak since the test would be stopped before the `ArrowArrayRelease` gets called.

zeroshade replied:
I wasn't as particularly worried about the memory leak since it was a test, but it's a good point.

davidwendt replied:
This pattern appears in other gtests as well.

zeroshade replied:
I've updated the other tests to use `EXPECT`

vyasr commented:
/ok to test

vyasr commented:
I've resolved the docs issue. We were getting a declaration (and definition) of ArrowSchema from nanoarrow's headers, whereas ArrowDeviceArray is only in the extension so we weren't seeing that one. There were namespacing issues around the lookup because of the name mangling that nanoarrow does when you specify `NANOARROW_NAMESPACE`, which again doesn't happen for ArrowDeviceArray for that reason. Rather than trying to fix this directly, though, I realized a simpler solution here would just be to link directly to the nanoarrow docs. I've added nanoarrow to the intersphinx lookup list, and now the docs link directly which is preferable in any case.

That does make me wonder, though, should we be mangling the name of the ArrowDeviceArray struct ourselves (it looks like nanoarrow changes the symbols to `cudf_*`)? We could assume that this isn't going to conflict with anything in the short term and just plan to switch over to using the nanoarrow device extension headers as soon as those stabilize. I'd be OK with that as long as we have a plan for that transition in a short to intermediate time frame.

vyasr commented:
/ok to test

vyasr commented:
/ok to test

zeroshade commented:
@vyasr Huzzah! all the CI passed!

vyasr commented on cpp/tests/interop/to_arrow_device_test.cpp:
```diff
@@ -0,0 +1,735 @@
+    if (expected->length > 0) {
+      EXPECT_EQ(expected->buffers[0], actual->buffers[0]);
+      if (schema_view.type == NANOARROW_TYPE_BOOL) {
+        // couldn't get thrust to like me, so let's just copy the data from device
+        // to host and then confirm the bytes are the same for the bool buffer
```
```suggestion
```

vyasr reviewed (changes requested):
This is looking really good. I think just one more iteration needed for me. Thanks for putting together a thorough test suite, that really helps.

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,680 @@
+  result->array              = private_data->parent;
+  result->array.private_data = private_data.release();
+  result->array.release      = &detail::ArrowDeviceArrayRelease;
+  return result;
```
@paleolimbot got me thinking about exactly what this API should look like in [this thread](https://github.com/apache/arrow-nanoarrow/pull/406#discussion_r1533130121). Right now, we're returning a `std::unique_ptr<ArrowDeviceArray>`, but we're effectively assuming that the consumer will take this and follow the C Data Interface rules. If the consumer does nothing at all with the output, then this would constitute a leak because the release hook would never have been called. Therefore, the way I see it we have two options that would be more internally consistent than what we do now:
- Construct this `unique_ptr` with a suitable deleter so that the `unique_ptr` handles the release properly even if the consumer does not. Then, the consumer is responsible for calling `release` on the returned `unique_ptr` to signal taking ownership of the array.
- Return a raw pointer to signal to the consumer that all memory management of this object is now their concern. This would be unsafe, but would at least be a little more honest than the current approach. IMO this is definitely the inferior of the two options, but I wanted to put it out there.

Does that assessment seem fair to you? Also definitely curious to hear Dewey's take here.

zeroshade replied:
@vyasr I chose to have it return the unique_ptr based on our conversations, but like @paleolimbot said, the common paradigm on the Arrow side is to accept a pointer output argument. Do you have a preference between setting the proper deleter vs taking an output pointer as an argument?

davidwendt replied:
I'm inclined to keep the unique_ptr and set the deleter. I think this is cleaner solution and more consistent with most libcudf APIs in general. I'd rather not start a pattern where pointers are provided for the API to fill in if possible. So given the choice I prefer a more C++ STL solution than a C one.

zeroshade replied:
Sure thing. I'll update to use a custom deleter with the unique ptrs then

vyasr commented:
@zeroshade one other question that I'm surprised we haven't discussed before (apologies if we have and I forgot), why are we choosing to return a single array with N children rather than N separate arrays? The current approach is perhaps more convenient in terms of setting up the arrow data structures and callbacks, but it will be less convenient to use, won't it?

vyasr commented:
#15370 exposed the fact that we're going to need the non-owning version of this API sooner rather than later, both for supporting a rewrite of the existing `to_arrow` host around the new `to_arrow_device` function and to support Python consumers. I'm going to restart that discussion in #14926 to avoid cluttering this PR, just wanted to let you know that I tried it out.

zeroshade commented on cpp/tests/interop/nanoarrow_utils.hpp:
```diff
@@ -0,0 +1,205 @@
+
+template <typename T>
+std::enable_if_t<cudf::is_fixed_width<T>() and !std::is_same_v<T, bool>, void> populate_from_col(
+  ArrowArray* arr, cudf::column_view view)
```
I think that would be more trouble than it's worth precisely because it would require us copying over the definition to make it useful. Not only that, but these functions are intended to be useful for children and recursive calling, if we expect it to be an `ArrowDeviceArray` we'd also have to pass the path down to the child we intend which could get extremely annoying . It's also much more convenient to have it just pass the `ArrowArray*` directly

vyasr replied:
Sounds right, thanks.

zeroshade commented on cpp/tests/interop/nanoarrow_utils.hpp:
added comments

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,680 @@
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<numeric::decimal128>(cudf::column&& column,
```
implemented with constexpr, let me know what you think

vyasr replied:
I think it's a marginal improvement. Not a big one, but enough that it's worth keeping. Thanks!

zeroshade commented:
> @zeroshade one other question that I'm surprised we haven't discussed before (apologies if we have and I forgot), why are we choosing to return a single array with N children rather than N separate arrays? The current approach is perhaps more convenient in terms of setting up the arrow data structures and callbacks, but it will be less convenient to use, won't it? IOW, why isn't this function taking in a cudf::column instead of a cudf::table?

The standard way to pass a table / record batch for Arrow across the C interface is as a struct column with the table columns making up the struct children. That said, I'll add an overload which takes in a cudf::column that just forwards it to the internal helpers.

vyasr commented:
> The standard way to pass a table / record batch for Arrow across the C interface is as a struct column with the table columns making up the struct children. That said, I'll add an overload which takes in a cudf::column that just forwards it to the internal helpers.

Got it, that makes sense. Thanks for bearing with me as I familiarize myself with best practices here. I do agree that we want the extra overload to exist for convenience.

Could you merge in branch-24.06 and retarget this PR? At this point we're into our code freeze period, so we'll have to merge this code into the next release. From my perspective, once that's done and we address the open question on whether to return a unique_ptr or require a user-provided output pointer as a parameter, I'm happy with this PR.

@davidwendt this is ready for another round of review from you. Thank you!

zeroshade commented:
@vyasr @davidwendt I've retargeted the PR, rebased, and updated it to use the custom deleter.

zeroshade commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -700,7 +706,10 @@ std::unique_ptr<ArrowDeviceArray> to_arrow_device(cudf::column&& col,
   ArrowArrayMove(tmp.get(), &private_data->parent);
-  auto result                = std::make_unique<ArrowDeviceArray>();
+  unique_device_array_t result(new ArrowDeviceArray, [](ArrowDeviceArray* arr) {
+    ArrowArrayRelease(&arr->array);
```
that's a good point, i'll update it

vyasr commented:
/ok to test

vyasr commented:
@zeroshade looks like some style checks failed, and also the tests need to be updated to not pass rvalues to the new span-based APIs since spans require lvalues.

zeroshade commented:
Hmm, if we can no longer pass rvalues to the API, what actually is the benefit of using the span over a const vector by reference?

vyasr commented:
/ok to test

vyasr commented:
/ok to test

vyasr commented:
/ok to test

zeroshade commented:
@vyasr looks like something failed with the conda download during the check-style. Not related to this PR as far as I can tell

davidwendt commented:
> @vyasr looks like something failed with the conda download during the check-style. Not related to this PR as far as I can tell

This should be fixed when https://github.com/rapidsai/cudf/pull/15403 is merged.

vyasr commented:
Yup unrelated. We pushed new CI images containing a conda update that removed a CLI option we relied on. We've temporarily downgraded and rebuilt our images to unblock things, but also the PR David linked is the long-term fix so that we can upgrade again.

> Hmm, if we can no longer pass rvalues to the API, what actually is the benefit of using the span over a const vector by reference?

That's a fair question. When the contained type is itself easily brace-initialized like in this case I believe you can get the same rvalue-like behavior by adding a second set of braces, which I believe will invoke the initializer list constructor of a `std::array` and then convert that to a span (the array's lifetime should persist for the length of the function call). [Here's a sample](https://godbolt.org/z/7vxrTo7W5). In general the benefit of a span is that you can use any container you want instead of being constrained to use a vector, but I assume you know that and you're asking for this specific API whether it's realistically important for users to be able to use other containers, and whether or not the typical usage pattern will just be via rvalues. I don't have a great answer for this API; it may not really be that important in practice and becomes just a matter of C++ hygiene.

vyasr commented:
/ok to test

vyasr reviewed (approved):
LGTM! Thanks for the hard work here @zeroshade! I'm approving assuming you don't plan to touch `from_arrow_device` in this PR (which I'm totally on board with, I think this PR is large and complicated enough as is).

zeroshade commented:
@vyasr yea, I plan on trying to tackle `from_arrow_device` in a follow-up PR

vyasr commented on cpp/src/interop/to_arrow_device.cu:
```diff
@@ -0,0 +1,722 @@
+  NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), NANOARROW_TYPE_STRING, column));
+
+  if (column.size() == 0) {
+    auto zero = std::make_unique<rmm::device_scalar<int32_t>>(0, stream, mr);
+    NANOARROW_RETURN_NOT_OK(set_buffer(std::move(zero), fixed_width_data_buffer_idx, tmp.get()));
```
Is this strictly necessary? Quoting from the spec:
> The buffer pointers MAY be null only in two situations:
>
>    for the null bitmap buffer, if [ArrowArray.null_count](https://arrow.apache.org/docs/format/> CDataInterface.html#c.ArrowArray.null_count) is 0;
>
>    **for any buffer, if the size in bytes of the corresponding buffer would be 0.**

https://arrow.apache.org/docs/format/CDataInterface.html#c.ArrowArray.buffers

zeroshade replied:
That's not for the null bitmap, it's for the offsets buffer. Many implementations expect there to be n+1 offsets, so a 0-length string array needs to have a single 0 value for its offset buffer

vyasr replied:
Sorry for any confusion, I had bolded the second point there, the one that says that you can leave the pointer null if the size of the corresponding buffer would be 0 for any buffer, not just the null mask. It sounds like from what you're saying that maybe many implementations are not allowing for that in this case, though, and they still dereference the offsets. Is that in violation of the spec, or am I reading it wrong?

vyasr replied:
Ah interesting OK thanks for the clarifications. From our perspective it would certainly be nicer if we could avoid the unnecessary allocation, and I would have expected the spec to promise that implementations wouldn't dereference the offsets of an empty string column. Could you add a comment here explaining why we need to allocate this scalar?

zeroshade replied:
Added the comment

vyasr replied:
Thanks!

vyasr commented:
Just a heads-up, NVIDIA employees in the US are off today and tomorrow so you probably won't see the next review until then.

vyasr commented:
/ok to test

davidwendt commented:
/ok to test

vyasr commented:
/merge

vyasr commented:
Thanks @zeroshade!
